base/set.jl

```diff
@@ -1,5 +1,42 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+"""
+    Set{T} <: AbstractSet{T}
+
+`Set`s are mutable containers that provide fast membership testing.
+
+`Set`s have efficient implementations of set operations such as `in`, `union` and `intersect`.
+Elements in a `Set` are unique, as determined by the elements' definitions of
+`isequal` and `hash`.
```

Comment:
I don't really get the addition of `hash` here. You could have a set with a bunch of elements that all hash to `1`, it would just be inefficient.

Reply:
The idea is to prevent a user overloading `isequal` without overloading `hash` (e,g, having a bunch of not-equal elements that hash to `1` is an implementation error). This is not legal according to `isequal` docs though, so it might be redundant to also state it here. I'll remove it.

Reply:
Here, the problem is kind of the opposite, i.e., having multiple `isequal` elements with different hashes. So, omitting `hash` may easily lead to hard-to-debug bugs, like
``` julia
julia> import Base: isequal, hash

julia> struct X; x; end

julia> isequal(X(0.0), X(0))
false

julia> isequal(a::X, b::X) = a.x == b.x;

julia> isequal(X(0.0), X(0))
true

julia> s = Set([X(0.0), X(0)]) # Incorrect
Set{X} with 2 elements:
  X(0)
  X(0.0)

julia> hash.([X(0.0), X(0)])
2-element Vector{UInt64}:
 0x29f6c96c4b9e8bb8
 0x0792f7046dd5cf45

julia> hash(x::X) = hash(x.x);

julia> s = Set([X(0.0), X(0)]) # Correct
Set{X} with 1 element:
  X(0)
```
Btw, `hash` is usually mention for other languages too. Just to name few: [C++](https://en.cppreference.com/w/cpp/container/unordered_set), [Rust](https://doc.rust-lang.org/std/collections/hash_set/struct.HashSet.html), [Java](https://docs.oracle.com/javase/8/docs/api/java/util/Set.html) ... unless the set is BST-based, of course.

Reply:
`isequal` is actually called to check `hash` collisions for `Dict` (which implements `Set` internally) in
https://github.com/JuliaLang/julia/blob/0062c2653dcddcbf9adb80d23b0ff5eef72f7802/base/dict.jl#L273

Reply:
Ah ok. So it does depend on both. 

Maybe the description should first say what the contract is (unique according to `isequal`, just like `unique`) and later say that the fast implementation needs `hash`? Uniqueness in the 1st paragraph, what's efficient in the second?

Reply:
Is that really necessary for the docstring, though? To me, mentioning `hash` seem like an optimization based on the implementation detail of Set, and that is usually outside the scope of a docstring.

Reply:
IMHO, yes! These two functions are entangled tightly. So, mentioning only one seems dangerous to me because it motivates the user to re-define `isequal` only, and it may lead to an incorrect/undefined behavior.